docs/proposals/dynamic-kubelet-configuration.md

```diff
@@ -0,0 +1,290 @@
+
+##### Reporting Configuration Status
+
+Succinct messages related to the state of `Node` configuration should be reported in a `NodeCondition`, in `status.conditions`. These should inform users which `ConfigMap` the Kubelet is using, and if the Kubelet has detected any issues with the configuration. The Kubelet should report this condition during startup, after attempting to validate configuration but before actually using it, so that the chance of a bad configuration inhibiting status reporting is minimized.
+
+All `NodeCondition`s contain the fields: `lastHeartbeatTime:Time`, `lastTransitionTime:Time`, `message:string`, `reason:string`, `status:string(True|False|Unknown)`, and `type:string`.
+
+These are some brief descriptions of how these fields should be interpreted for node-configuration related conditions:
+- `lastHeartbeatTime`: The last time the Kubelet updated the condition. The Kubelet will typically do this whenever it is restarted, because that is when configuration changes occur. The Kubelet will update this on restart regardless of whether the configuration, or the condition, has changed.
+- `lastTransitionTime`: The last time this condition changed. The Kubelet will not update this unless it intends to set a different condition than is currently set.
+- `message`: A report on the Think of this as the "effect" of the `reason`.
+- `reason`: Think of this as the "cause" of the `message`.
+- `status`: `True` if the currently set configuration is considered OK, `False` otherwise. `Unknown` should not be used in these condition messages. One *might* say that a config in its trial period has "unknown" goodness, but the `status` should still be `True` because the Kubelet is treating it as OK.
+- `type`: `ConfigOK` will always be used for these conditions.
+
+The following list of example conditions, sans `type`, `lastHeartbeatTime`, and `lastTransitionTime`, can be used to get a feel for the relationship between `message`, `reason`, and `status`:
+
+Config still in trial period:
+```
+message: "using current: namespace/name"
+reason: "in trial period: namespace/name"
+status: "True"
+```
+
+Config passed trial period:
+```
+message: "using current: namespace/name"
+reason: "passed trial period: namespace/name"
+status: "True"
+```
+
+No remote config specified:
+```
+message: "using local init config"
+reason: "this Node's spec.config was empty"
+status: "True"
+```
+
+If `Node.spec.config` refers to an object which is not a `ConfigMap`, we treat similarly to an empty `spec.config`, but report status `False`, as this is likely an error:
```

Comment:
OTOH, we could default to lkg instead of init here, which would make an accidentally malformed rollout less disruptive